Fill in kubelet metrics API

timstclair opened the pull request:
This is a first pass at implementing the new metrics API group and types described in https://github.com/kubernetes/kubernetes/pull/15862

The bulk of the API added here is copied verbatim from the [cAdvisor v2](https://github.com/google/cadvisor/blob/master/info/v2/container.go) (and v1 where relevant) info API. I'm hoping we can leverage (in a follow-up PR) the generated conversion infrastructure to convert the cAdvisor API for us, which will greatly decrease the overhead of maintaining our own version of the API.

**Discussion points:**
1. Should the `Raw{Node,Pod,Container}` objects have an `ObjectMeta`, `ObjectRefrence`, or both?
   - ~~The [proposal](https://github.com/kubernetes/kubernetes/pull/15862) suggested embedding ObjectMeta in the top level (`RawNode`, `RawPod`, and `RawContainer`) metrics, but reading the descriptions of [ObjectMeta](https://github.com/kubernetes/kubernetes/blob/9c4424f0bd55b8f2fb4610a4616fbd5cd30abc9a/pkg/api/types.go#L60) and [ObjectReference](https://github.com/kubernetes/kubernetes/blob/9c4424f0bd55b8f2fb4610a4616fbd5cd30abc9a/pkg/api/types.go#L1740) it sounds like `ObjectReference` would be more appropriate here. The reasoning is that 1) metrics are not persisted objects and don't have an identity (except for possibly samples), and 2) the metrics object _refers_ to the API object (e.g. Pod) as opposed to _being_ the object.~~
   - After studying API conventions & offline discussions, I think using `ListMeta` on the top-level with references to the Pod/Node/Container is the best fit. For containers, I've restricted the reference to just a Name for data size, and since they are not top-level objects in this API.
2. How do we want to handle cases where the cAdvisor API deviates from our [API conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md)?
   - Decision: MInimize changes for v1alpha1 API. Prior to advancing to beta status, the API needs to be made conformant, and to get there we'd like to create a new cAdvisor API which conforms to our API conventions.

@vishh @lavalamp @bgrant0607 @kubernetes/goog-node

timstclair commented:
BTW - I know the generated types don't build, I'm still investigating why, but I'm still interested in feedback on the posted questions and API.

timstclair commented:
FYI, fixed codecgen (problem was a field of type `map[string][]MetricVal`)

tallclair committed:
WIP: Adding cAdvisor types to metrics API group

timstclair commented:
I've removed everything from this PR except for types.go so we can focus the review attention on the API. Once we have settled on the API, I'll make the v1alpha1 version and regenerate all the necessary files.

Note that almost all of this API is copied from the cAdvisor [v2](https://github.com/google/cadvisor/blob/master/info/v2/container.go) (and in a couple cases [v1](https://github.com/google/cadvisor/blob/master/info/v1/container.go)) API. Changes include:
  - Adding missing json annotations, renaming json fields to match our conventions, and adding `omitempty` where necessary
  - Adding missing documentation on several fields, along with links to kernel documentaiton
  - Changing the type of CustomMetrics to a list of named values, since the map type was problematic for ugorji

cc/ @vishh

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
 
-import "k8s.io/kubernetes/pkg/api/unversioned"
+import (
+	"time"
```
Use the time in unversioned, it will marshal correctly.

timstclair replied:
Done.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
 	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata, applying to the lists of stats.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:",inline"`
```
I need to double-check, but I think we marshal this as "metadata"

lavalamp replied:
You'll need this even for alpha.

timstclair replied:
Done

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	"time"
 
-// Placeholder top-level node resource metrics.
+	"k8s.io/kubernetes/pkg/api"
```
Definitely can't import pkg/api here. Get common things from unversioned or copy them here.

lavalamp replied:
You'll need this even for alpha.

timstclair replied:
Done

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:",inline"`
+	// Reference to the measured Node.
+	NodeRef api.ObjectReference `json:"nodeRef,omitempty"`
```
Yeah, you can't directly use anything from pkg/api.

lavalamp replied:
You could consider just using NodeName here, too. Since it's not namespaced.

lavalamp replied:
You'll need this even for alpha.

timstclair replied:
Done

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// Reference to the measured container.
+	Name string `json:"name,omitempty"`
+	// Describes the container resources.
+	Spec ContainerSpec `json:"spec,omitempty"`
```
Is this really a Spec, in the sense we normally use that term? I.e., would a user POST/PUT/PATCH and change or set this data? If not, I think this is actually Status, not Spec.

lavalamp replied:
Maybe this is more like a schema...

timstclair replied:
Ack.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+type CpuSpec struct {
+	// Requested cpu shares. Default is 1024.
+	// Units: milli-cpus.
+	Limit uint64 `json:"limit,omitempty"`
```
Please see pkg/api/resource.Quantity; I think we should use it everywhere that is appropriate.

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+
+type CpuSpec struct {
+	// Requested cpu shares. Default is 1024.
+	// Units: milli-cpus.
```
I would understand the first line of the comment to mean that the units are *1/1024ths* of a *core*, not *1/1000ths* of the CPU, but I also think the API should return milli-cpus, to be consistent with our existing api.

I'm worried in general about getting units mixed up.

timstclair replied:
@vishh Do you know the answer to this? It doesn't even look like this struct is ever used?

vishh replied:
This field mirrors the kernel interface. Needless to say the kernel prefers powers of 2 over decimal values for limits. 
@timstclair: Kubelet partitions cpu based on a fixed number of shares. May be the kubelet can override this value with an appropriate CPU limit instead?
For example, 1024 shares would translate to 1 cpu or 1000 milli cpus.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+type MemorySpec struct {
+	// The amount of memory requested. Default is unlimited (-1).
+	// Units: bytes.
+	Limit uint64 `json:"limit,omitempty"`
```
resource.Quantity here, too.

timstclair replied:
Ack

vishh replied:
If we are using quantity here, why not use it in all places?

timstclair replied:
I think eventually we'll want something like quantity in all places, but the decision was to leave it unchanged for this version of the api.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// Metadata labels associated with this container.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	HasCpu bool    `json:"hasCpu,omitempty"`
```
Not a fan of "HasX" fields. Can you just make the CPU field a pointer?

timstclair replied:
Agreed, will fix in future version (Ack)

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	HasCpu bool    `json:"hasCpu,omitempty"`
+	Cpu    CpuSpec `json:"cpu,omitempty"`
```
From go naming convention point of view, this should be:
```
CPU CPUSpec `json:"cpu,omitempty"`
```

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	Namespace string `json:"namespace,omitempty"`
+
+	// Metadata labels associated with this container.
+	Labels map[string]string `json:"labels,omitempty"`
```
Aren't our labels on pods, not containers? I don't think this is the right place for this.

timstclair replied:
As with namespace, these aren't kubernetes labels (they're docker labels). Added a comment to that effect.

lavalamp commented:
More later.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	CpuInst *CpuInstStats `json:"cpuInst,omitempty"`
+	// Disk IO statistics
+	HasDiskIo bool        `json:"hasDiskIo,omitempty"`
+	DiskIo    DiskIoStats `json:"diskIo,omitempty"`
```
DiskIO/diskIO

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+}
+
+// Instantaneous CPU stats
+type CpuInstStats struct {
```
No abbreviations in the API; CPUInstantaneousStats.

(We do make a small handful of exceptions, "Stats" for statistics is probably OK.)

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+// CPU usage time statistics.
+type CpuInstUsage struct {
+	// Total CPU usage.
+	// Units: nanocores per second
```
Can we combine the instantaneous metric and the below struct by adding a Period field? That is, all of the values become a duration, and we add a "period"/"duration" field, so that you can just divide if you need to.

Also shouldn't this unit be something like nanocore-seconds per second?

lavalamp replied:
(I don't believe in instantaneous measurements, that would imply you divided by zero. The measurement period must have been non-zero.)

timstclair replied:
Yeah, these units are confusing to me. @vishh do you know the answer to this?

vishh replied:
If a process used 1 cpu in the last one second, then its instantaneous usage measured every second is `10^9 nanocores/s`. 
If a process used 2 cpus in last 5 seconds, then its instantaneous usage measured every 5 seconds is `4*10^8 nanocores/s`. 

Does this make sense? The units is nanocores to provide the best possible precision and avoid divide by zero issues.

timstclair replied:
I think Daniel's points are valid. It doesn't make sense to use 1 cpu _in_ one second, you would use 1 cpu _for_ 1 second (1 cpu-second), so then the "instantaneous" (which is really a 10-second rolling average) usage would be N cpu-seconds / second, which I think is N*10^9 nanocore-seconds / second? (or does a nanocore include seconds?)

I think we've concluded that changing the units is out-of-scope for this PR, but I do want to at least get the units documented correctly.

vishh replied:
Yeah. That makes sense. Maybe `cpu-nanoseconds / second`?

On Thu, Nov 12, 2015 at 1:36 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44718079>
> :
>
> > +	TxPackets uint64 `json:"txPackets,omitempty"`
> > +	// Cumulative count of transmit errors encountered.
> > +	TxErrors uint64 `json:"txErrors,omitempty"`
> > +	// Cumulative count of packets dropped while transmitting.
> > +	TxDropped uint64 `json:"txDropped,omitempty"`
> > +}
> > +
> > +// Instantaneous CPU stats
> > +type CpuInstStats struct {
> > +	Usage CpuInstUsage `json:"usage,omitempty"`
> > +}
> > +
> > +// CPU usage time statistics.
> > +type CpuInstUsage struct {
> > +	// Total CPU usage.
> > +	// Units: nanocores per second
>
> I think Daniel's points are valid. It doesn't make sense to use 1 cpu *in*
> one second, you would use 1 cpu *for* 1 second (1 cpu-second), so then
> the "instantaneous" (which is really a 10-second rolling average) usage
> would be N cpu-seconds / second, which I think is N*10^9 nanocore-seconds /
> second? (or does a nanocore include seconds?)
>
> I think we've concluded that changing the units is out-of-scope for this
> PR, but I do want to at least get the units documented correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44718079>.
>

lavalamp replied:
`core-nanoseconds/second`. `CPU` is ambiguous.

On Thu, Nov 12, 2015 at 2:05 PM, Vish Kannan <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44721928>
> :
>
> > +	TxPackets uint64 `json:"txPackets,omitempty"`
> > +	// Cumulative count of transmit errors encountered.
> > +	TxErrors uint64 `json:"txErrors,omitempty"`
> > +	// Cumulative count of packets dropped while transmitting.
> > +	TxDropped uint64 `json:"txDropped,omitempty"`
> > +}
> > +
> > +// Instantaneous CPU stats
> > +type CpuInstStats struct {
> > +	Usage CpuInstUsage `json:"usage,omitempty"`
> > +}
> > +
> > +// CPU usage time statistics.
> > +type CpuInstUsage struct {
> > +	// Total CPU usage.
> > +	// Units: nanocores per second
>
> Yeah. That makes sense. Maybe `cpu-nanoseconds / second`?
> … <#150fdbbcbd3a206d_>
> On Thu, Nov 12, 2015 at 1:36 PM, Tim St. Clair <notifications@github.com>
> wrote: In pkg/apis/metrics/types.go <#16458 (comment)
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44718079>>
> : > + TxPackets uint64 `json:"txPackets,omitempty"` > + // Cumulative count
> of transmit errors encountered. > + TxErrors uint64
> `json:"txErrors,omitempty"` > + // Cumulative count of packets dropped
> while transmitting. > + TxDropped uint64 `json:"txDropped,omitempty"` > +}
> > + > +// Instantaneous CPU stats > +type CpuInstStats struct { > + Usage
> CpuInstUsage `json:"usage,omitempty"` > +} > + > +// CPU usage time
> statistics. > +type CpuInstUsage struct { > + // Total CPU usage. > + //
> Units: nanocores per second I think Daniel's points are valid. It doesn't
> make sense to use 1 cpu *in* one second, you would use 1 cpu *for* 1 second
> (1 cpu-second), so then the "instantaneous" (which is really a 10-second
> rolling average) usage would be N cpu-seconds / second, which I think is
> N*10^9 nanocore-seconds / second? (or does a nanocore include seconds?) I
> think we've concluded that changing the units is out-of-scope for this PR,
> but I do want to at least get the units documented correctly. — Reply to
> this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/16458/files#r44718079>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44721928>.
>

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+
+	// Per CPU/core usage of the container.
+	// Unit: nanocores per second
+	PerCpu []uint64 `json:"perCpu,omitempty"`
```
Are there any guarantees about this array? Like can I expect there to be an entry for every CPU in the system, and if so, what order are the CPUs in?

vishh replied:
Usage on all CPUs is expected. It should be in increasing order - CPU 0 through CPU MAX.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+// CPU usage time statistics.
+type CpuUsage struct {
+	// Total CPU usage.
+	// Units: nanoseconds
```
nanocore-seconds?

(Is nanocores really necessary? Why not micro or event milli core?)

vishh replied:
This is not in cores. This is in nanoseconds. If a cgroup uses 1 cpu continuously for 5 seconds, this value will be `5*10^9` nanoseconds. And this value is cumulative.

timstclair replied:
But if it uses 2 cpus for 5 seconds, isn't it `10*10^9`? So shouldn't it be core-nanoseconds == nanocore-seconds?

lavalamp replied:
The unit you just described is "core-nanoseconds", not "nanoseconds"

On Thu, Nov 12, 2015 at 11:08 AM, Vish Kannan <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44699843>
> :
>
> > +	// Unit: nanocores per second
> > +	PerCpu []uint64 `json:"perCpu,omitempty"`
> > +
> > +	// Time spent in user space.
> > +	// Unit: nanocores per second
> > +	User uint64 `json:"user,omitempty"`
> > +
> > +	// Time spent in kernel space.
> > +	// Unit: nanocores per second
> > +	System uint64 `json:"system,omitempty"`
> > +}
> > +
> > +// CPU usage time statistics.
> > +type CpuUsage struct {
> > +	// Total CPU usage.
> > +	// Units: nanoseconds
>
> This is not in cores. This is in nanoseconds. If a cgroup uses 1 cpu
> continuously for 5 seconds, this value will be 5*10^9 nanoseconds. And
> this value is cumulative.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44699843>.
>

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+// All CPU usage metrics are cumulative from the creation of the container
+type CpuStats struct {
+	Usage CpuUsage `json:"usage,omitempty"`
+	// Smoothed average of number of runnable threads x 1000.
```
I don't really follow this. What do runnable threads have to do with a load average? What does load average mean? Why is 10 seconds hard coded?

timstclair replied:
10 seconds is the period that samples are gathered at. As to why it's hard coded, good question :)

vishh replied:
These stats are relatively expensive to gather. So they are calculated every ten seconds. Load average can be interpreted as the CPU load that a cgroup is experiencing. If the load is high, then the cgroup needs more CPUs.

lavalamp replied:
Maybe the comment can be updated to reflect that. Can we either ban the
word "CPU" in this file (use "core" instead), or explicitly define it?

On Thu, Nov 12, 2015 at 11:10 AM, Vish Kannan <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44700107>
> :
>
> > +	// Unit: nanoseconds.
> > +	PerCpu []uint64 `json:"perCpu,omitempty"`
> > +
> > +	// Time spent in user space.
> > +	// Unit: nanoseconds
> > +	User uint64 `json:"user,omitempty"`
> > +
> > +	// Time spent in kernel space.
> > +	// Unit: nanoseconds
> > +	System uint64 `json:"system,omitempty"`
> > +}
> > +
> > +// All CPU usage metrics are cumulative from the creation of the container
> > +type CpuStats struct {
> > +	Usage CpuUsage `json:"usage,omitempty"`
> > +	// Smoothed average of number of runnable threads x 1000.
>
> These stats are relatively expensive to gather. So they are calculated
> every ten seconds. Load average can be interpreted as the CPU load that a
> cgroup is experiencing. If the load is high, then the cgroup needs more
> CPUs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44700107>.
>

timstclair replied:
Updated comment. Is the difference between this and LoadStats.NrRunning just that this is smoothed over the last 10 seconds, and NrRunning is just the number counted at the instant the measurement was taken?

timstclair replied:
> Maybe the comment can be updated to reflect that. Can we either ban the
word "CPU" in this file (use "core" instead), or explicitly define it?

I think CPU should refer to the abstract resource (e.g. CPU stats), while core refers to the actual measurement (core is a unit). I updated a couple of comments to reflect that. Does it make sense?

vishh replied:
Higher Load implies that processes in a container are waiting longer to get
access to Cpu. 100% CPU utilization does not necessarily imply that the
`Load` for the container is higher.

On Thu, Nov 12, 2015 at 3:46 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44732760>
> :
>
> > +	// Unit: nanoseconds.
> > +	PerCpu []uint64 `json:"perCpu,omitempty"`
> > +
> > +	// Time spent in user space.
> > +	// Unit: nanoseconds
> > +	User uint64 `json:"user,omitempty"`
> > +
> > +	// Time spent in kernel space.
> > +	// Unit: nanoseconds
> > +	System uint64 `json:"system,omitempty"`
> > +}
> > +
> > +// All CPU usage metrics are cumulative from the creation of the container
> > +type CpuStats struct {
> > +	Usage CpuUsage `json:"usage,omitempty"`
> > +	// Smoothed average of number of runnable threads x 1000.
>
> Updated comment. Is the difference between this and LoadStats.NrRunning
> just that this is smoothed over the last 10 seconds, and NrRunning is just
> the number counted at the instant the measurement was taken?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44732760>.
>

vishh replied:
`Core` can mean different things. Whereas `Cpu` is unambiguous.

On Thu, Nov 12, 2015 at 3:50 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44733131>
> :
>
> > +	// Unit: nanoseconds.
> > +	PerCpu []uint64 `json:"perCpu,omitempty"`
> > +
> > +	// Time spent in user space.
> > +	// Unit: nanoseconds
> > +	User uint64 `json:"user,omitempty"`
> > +
> > +	// Time spent in kernel space.
> > +	// Unit: nanoseconds
> > +	System uint64 `json:"system,omitempty"`
> > +}
> > +
> > +// All CPU usage metrics are cumulative from the creation of the container
> > +type CpuStats struct {
> > +	Usage CpuUsage `json:"usage,omitempty"`
> > +	// Smoothed average of number of runnable threads x 1000.
>
> Maybe the comment can be updated to reflect that. Can we either ban the
> word "CPU" in this file (use "core" instead), or explicitly define it?
>
> I think CPU should refer to the abstract resource (e.g. CPU stats), while
> core refers to the actual measurement (core is a unit). I updated a couple
> of comments to reflect that. Does it make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44733131>.
>

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+// Disk IO stats, as reported by the cgroup block io controller.
+// See https://www.kernel.org/doc/Documentation/cgroups/blkio-controller.txt
+type DiskIoStats struct {
+	IoServiceBytes []PerDiskStats `json:"ioServiceBytes,omitempty"`
```
IO

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	WorkingSet uint64 `json:"workingSet,omitempty"`
+
+	// Cumulative number of times that a usage counter hit its limit
+	Failcnt uint64 `json:"failcnt,omitempty"`
```
FailureCount

timstclair replied:
Ack

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+
+type MemoryStatsMemoryData struct {
+	// Cumulative number of minor page faults.
+	Pgfault uint64 `json:"pgfault,omitempty"`
```
PageFault/Pagefault-- I think the former, not 100% sure.

timstclair replied:
A lot of the field names in these structs are copied from the stat names reported by the linux kernel. On the one hand, it makes it clear what the fields are reporting. On the other hand, this could be viewed as an implementation detail which should be masked.

lavalamp replied:
We probably shouldn't rule out running on other platforms (windows) in the future.

vishh replied:
These are linux APIs. If windows shows up in the future, we might have to expose a completely different set of metrics. Users of these metrics will have to abstract out the underlying OS.

vishh replied:
Even though they mirror Linux APIs, since these are read-only, I think its OK to make them more user-friendly.
It will avoid having to look up kernel documentation.

timstclair replied:
Ack (will address in future version)

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	Pgmajfault uint64 `json:"pgmajfault,omitempty"`
+}
+
+type FsStats struct {
```
FilesystemStats

timstclair replied:
Ack.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+// This mirrors kernel internal structure.
+type LoadStats struct {
+	// Number of sleeping tasks.
+	NrSleeping uint64 `json:"nrSleeping,omitempty"`
```
Number

timstclair replied:
Ack (I prefer Count)

lavalamp replied:
Count works for me too, I almost suggested it.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+
+const (
+	IntType   DataType = "int"
+	FloatType DataType = "float"
```
I suggest looking at our "IntOrString" type and doing something similar.

lavalamp commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// Network stats by interface.
+	Interfaces []InterfaceStats `json:"interfaces,omitempty"`
+	// TCP connection stats (Established, Listen...)
+	Tcp TcpStat `json:"tcp,omitempty"`
```
TCP

timstclair replied:
Ack

lavalamp commented:
I think my big worry with this is the units, other comments more or less minor. If this were just straight Go code, I would say to make a different type for every unit, and define the allowed conversions in functions-- that way, the compiler will catch people that are doing bad things with units. I think that may still be my recommendation.

lavalamp commented:
After chatting with Tim, I think I'll withdraw most of my comments until we want to take this to beta. I'll mark a few that should still be fixed.

tallclair committed:
@lavalamp feedback

timstclair commented:
Thanks for the review. I addressed the comments that need to be addressed for v1alpha1. Also I want to fix the documentation (since those won't complicate things), but I'm unsure about the answers to some of your questions. I'll try to track down the answers if @vishh doesn't know.

lavalamp commented:
LGTM for an alpha API

bgrant0607 commented:
Quick question: Why is ObjectReference being moved to unversioned.

vishh commented:
@timstclair: Once we mature this API to be of Beta quality, let's try to make cAdvisor expose this API instead of its existing API. That will help avoid (or reduce) the transcoding logic in Kubelet.

timstclair commented:
@bgrant0607  - I think the reasoning is we want to avoid cross-version dependencies between APIs. This way when we get to a v2 API we aren't forced to either update all the API groups or live with API groups referencing deprecated versions. (@lavalamp anything to add?)

@vishh - Thanks for the responses. I'm still a little unsure of CPU units though.

lavalamp commented:
Because the internal api.ObjectReference should not be serialized.

On Thu, Nov 12, 2015 at 10:57 AM, Brian Grant <notifications@github.com>
wrote:

> Quick question: Why is ObjectReference being moved to unversioned.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458#issuecomment-156201229>
> .
>

tallclair committed:
Updated type comments

tallclair committed:
Update generated files

timstclair commented:
I think this discussion is wrapping up so I've copied the `metrics/types.go` changes to `metrics/v1alpha1/types.go`, updated `metrics/{,v1alpha1/}register.go`, and updated generated files.

I've removed the "WIP" label, so this is ready for final review.

timstclair commented:
It turns out we need a beta API for 1.2, which means we need to rewrite this API to conform to our standards and the data structures we want for the longer term. I'll open a new PR once those changes are made.

swagiaal commented:
cc @lberk
